sound/soc/soc-pcm.c

```diff
@@ -2015,76 +2017,141 @@ int dpcm_be_dai_trigger(struct snd_soc_pcm_runtime *fe, int stream,
 
 		switch (cmd) {
 		case SNDRV_PCM_TRIGGER_START:
+			spin_lock_irqsave(&fe->card->dpcm_lock, flags);
 			if ((be->dpcm[stream].state != SND_SOC_DPCM_STATE_PREPARE) &&
 			    (be->dpcm[stream].state != SND_SOC_DPCM_STATE_STOP) &&
-			    (be->dpcm[stream].state != SND_SOC_DPCM_STATE_PAUSED))
+			    (be->dpcm[stream].state != SND_SOC_DPCM_STATE_PAUSED)) {
```

Comment:
> @plbossart just curious, this patch wouldnt really prevent the BE from getting triggered twice would it? It would just stagger the multiple triggers right?

That's right, this only fixes a consistency problem, we had a race condition with multiple triggers where the state was read before being modified.

The next patch adds restrictions to limit the number of triggers.

Reply:
> I also think if we're going to set the state with the lock held, then all reads should also be with the lock held no?

That's what I am doing in this function, isn't it?
Are you asking a generic question or do you see a problem i missed?

Reply:
in this function yes, but do we need to care about others too? Or do we only care about a race between 2 triggers?

Reply:
I looked at the code and only saw a couple of cases where the spinlock is not used, e.g.

dpcm_be_dai_startup()
dpcm_be_dai_stop()
pcm_be_dai_hw_free()
dpcm_be_dai_hw_free()

but for all of them there is a mutex anyways in the caller, e.g.

````c
static int dpcm_fe_dai_hw_params(struct snd_pcm_substream *substream,
				 struct snd_pcm_hw_params *params)
{
	struct snd_soc_pcm_runtime *fe = asoc_substream_to_rtd(substream);
	int ret, stream = substream->stream;

	mutex_lock_nested(&fe->card->mutex, SND_SOC_CARD_CLASS_RUNTIME);
````

So I would assume the risk is only between triggers where the mutex is not used.

Reply:
fair enough!